In [1]:
#!pip install gdown

In [2]:
#!gdown https://drive.google.com/uc?id=1MhctnTnP8PnBYKdlAtY7y_GVqXZzV2Jz

In [3]:
import anndata as ad
import pandas as pd

In [4]:
pd.options.display.max_columns = 100

## Check L1000 structure

In [5]:
adata_l1000 = ad.read_h5ad('./l1000_standardized.h5ad')

In [6]:
adata_sci = ad.read_h5ad('./data/sciplex/pseudobulk/subsample/srivatsan20_sciplex3_subsample.h5ad')

In [7]:
adata_l1000

AnnData object with n_obs × n_vars = 1994606 × 978
    obs: 'bead_batch', 'nearest_dose', 'pert_dose', 'pert_dose_unit', 'pert_idose', 'pert_time', 'pert_itime', 'pert_time_unit', 'cell_mfc_name', 'pert_mfc_id', 'rna_plate', 'rna_well', 'count_mean', 'count_cv', 'qc_f_logp', 'qc_iqr', 'qc_slope', 'pert_id', 'pert_type', 'qc_pass', 'dyn_range', 'inv_level_10', 'build_name', 'failure_mode', 'project_code', 'cmap_name', 'is_control', 'mw_g_per_mol', 'inchi_key', 'compound_aliases', 'pubchem_cid', 'alternative_alias', 'dose_val', 'pert_dose_uM', 'dose_bin_uM', 'time_val', 'time_unit', 'pert_time_h', 'pert_time_bin', 'plate', 'well', 'cell_type', 'perturbagen', 'sample_id', 'perturbation_label', 'psbulk_cells'
    var: 'symbol', 'pr_gene_symbol', 'pr_is_lm'

In [8]:
adata_sci

AnnData object with n_obs × n_vars = 1459 × 58302
    obs: 'plate', 'well', 'cell_type', 'perturbagen', 'pert_type', 'is_control', 'pert_dose_uM', 'pert_time_h', 'suspension_type', 'tissue', 'tissue_type', 'disease', 'library', 'stimulation', 'guide', 'dataset', 'assay', 'development_stage', 'organism', 'sex', 'self_reported_ethnicity', 'pubchem_cid', 'psbulk_cells', 'psbulk_counts'
    var: 'symbol'
    layers: 'psbulk_props'

### .obs

In [9]:
# Excessive columns in comparison to our reference structure in adata_sci:
adata_l1000.obs[list(set(adata_l1000.obs.columns) - set(adata_sci.obs.columns))]

,rna_well,cmap_name,inv_level_10,build_name,time_unit,dose_val,cell_mfc_name,count_cv,pert_time_bin,pert_id,sample_id,project_code,pert_dose_unit,qc_iqr,alternative_alias,pert_dose,rna_plate,bead_batch,count_mean,time_val,qc_slope,mw_g_per_mol,pert_mfc_id,dyn_range,qc_pass,pert_time_unit,qc_f_logp,pert_itime,compound_aliases,dose_bin_uM,perturbation_label,nearest_dose,pert_time,failure_mode,pert_idose,inchi_key
0,D10,LACZ,1462.0,NaN,h,0.00000,U2OS,18,96h,TRCN0000072237,TAK004_U2OS_96H_X2_B10_DUO52HI53LO_D10_TRCN000...,TAK,uL,14.98,NaN,1.00000,TAK004_U2OS_96H_X1,b10,67,96.0,67,NaN,TRCN0000072237,4.73906,0.0,h,5.7,96 h,NaN,≤0.1 µM,TRCN0000072237,NaN,96.0,inv_level_10,1 uL,NaN
1,D17,BRD-K79781870,1558.0,NaN,h,5.33000,A375,26,24h,BRD-K79781870,DOS043_A375_24H_X1_F3B5_DUO52HI53LO_D17_BRD-K7...,DOS,uM,15.87,BRD-K79781870,5.33000,DOS043_A375_24H_X1,f3b5,58,24.0,68,536.282112,BRD-K79781870,6.78867,0.0,h,6.3,24 h,NaN,1–10 µM,BRD-K79781870,6.660,24.0,"inv_level_10,qc_iqr",6.66 uM,VSPYGNDYHLSTDT-UOBPVEMNSA-N
2,P11,tanespimycin,2301.0,NaN,h,0.00150,A549,20,6h,BRD-A61304759,HOG001_A549_6H_X4_F1B10_P11_BRD-A61304759_A549,HOG,uM,9.95,tanespimycin,0.00150,HOG001_A549_6H_X4,f1b10,65,6.0,62,585.305015,BRD-A61304759,12.47150,1.0,h,9.8,6 h,NaN,≤0.1 µM,BRD-A61304759,0.002,6.0,NaN,0.002 uM,AYUNIORJHRXIBJ-ZGQRYRSUSA-N
3,A12,LUCIFERASE,1511.0,NaN,h,0.00000,VCAP,14,120h,TRCN0000072261,ERG013_VCAP_120H_X1_B11_A12_TRCN0000072261_VCAP,ERG,uL,22.92,NaN,20.00000,ERG013_VCAP_120H_X1,b11,108,120.0,70,NaN,TRCN0000072261,3.31360,0.0,h,5.9,120 h,NaN,≤0.1 µM,TRCN0000072261,NaN,120.0,"inv_level_10,qc_iqr,dyn_range",20 uL,NaN
4,O04,GFP,3344.0,NaN,h,0.00000,HEK293T,15,48h,GFP,HSF043_HEK293T_48H_X1_B12_O04_GFP_HEK293T,HSF,ng,11.23,NaN,150.00000,HSF043_HEK293T_48H_X1,b12,91,48.0,63,NaN,GFP,6.44316,1.0,h,6.5,48 h,NaN,≤0.1 µM,GFP,NaN,48.0,NaN,150 ng,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1994601,N04,PF-05212384,4804.0,NaN,h,0.15625,BJAB,13,4h,BRD-K07955840,AICHI002_BJAB_4H_X1.A2_B40_N04_BRD-K07955840_BJAB,AICHI,uM,14.23,NaN,0.15625,AICHI002_BJAB_4H_X1.A2,b40,113,4.0,54,615.328151,BRD-K07955840,14.02630,1.0,h,5.9,4 h,NaN,0.1–1 µM,BRD-K07955840,0.125,4.0,NaN,0.125 uM,DWZAEMINVBZMHQ-UHFFFAOYSA-N
1994602,H10,apicidin,3304.0,NaN,h,0.15625,MCF7,16,24h,BRD-K64606589,HDAC001_MCF7_24H_X1_F1B3_DUO52HI53LO_H10_BRD-K...,HDAC,uM,13.17,apicidin,0.15625,HDAC001_MCF7_24H_X1,f1b3,103,24.0,68,623.368284,BRD-K64606589,5.14642,1.0,h,8.6,24 h,NaN,0.1–1 µM,BRD-K64606589,0.125,24.0,NaN,0.125 uM,JWOGUUIOCYMBPV-VWIQTCEGSA-N
1994603,N04,BRD-K61894884,2986.0,NaN,h,0.15625,PC3,19,24h,BRD-K61894884,HDAC001_PC3_24H_X1_F1B3_DUO52HI53LO_N04_BRD-K6...,HDAC,uM,11.80,BRD-K61894884,0.15625,HDAC001_PC3_24H_X1,f1b3,78,24.0,67,NaN,BRD-K61894884,8.70554,1.0,h,9.0,24 h,NaN,0.1–1 µM,BRD-K61894884,0.125,24.0,NaN,0.125 uM,NaN
1994604,L07,enalapril,2378.0,NaN,h,0.12350,MCF7,17,24h,BRD-K57545991,RAD001_MCF7_24H_X3_F1B5_DUO52HI53LO_L07_BRD-K5...,RAD,uM,16.85,enalapril,0.12350,RAD001_MCF7_24H_X3,f1b5,66,24.0,67,376.199822,BRD-K57545991,6.05089,0.0,h,7.6,24 h,NaN,0.1–1 µM,BRD-K57545991,0.125,24.0,qc_iqr,0.125 uM,GBXSMTUPTTWBMN-XIRDDKMYSA-N


In [10]:
# No columns in comparison to our reference structure in adata_sci:
adata_sci.obs[list(set(adata_sci.obs.columns) - set(adata_l1000.obs.columns))]

,development_stage,suspension_type,tissue_type,library,guide,sex,dataset,assay,stimulation,psbulk_counts,disease,tissue,self_reported_ethnicity,organism
sample_id,,,,,,,,,,,,,,
plate10_plate4_C7_panobinostat_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,832,lung adenocarcinoma,lung,European,human
plate10_plate4_H12_PCI-34051_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,746,lung adenocarcinoma,lung,European,human
plate11_plate5_B12_A-366_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,704,lung adenocarcinoma,lung,European,human
plate11_plate5_B8_AR-42_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,1110,lung adenocarcinoma,lung,European,human
plate11_plate5_B9_AMG-900_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,1283,lung adenocarcinoma,lung,European,human
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
plate9_plate3_A4_tazemetostat_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,755,lung adenocarcinoma,lung,European,human
plate9_plate3_A7_SRT2104_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,1169,lung adenocarcinoma,lung,European,human
plate9_plate3_F1_Divalproex_CVCL_0023,unknown,nucleus,cell culture,NaN,NaN,male,srivatsan20_sciplex3,sci-Plex,NaN,2164,lung adenocarcinoma,lung,European,human


### .var

In [11]:
# Excessive columns in comparison to our reference structure in adata_sci:
adata_l1000.var[list(set(adata_l1000.var.columns) - set(adata_sci.var.columns))]

,pr_is_lm,pr_gene_symbol
pr_gene_symbol,,
GNPDA1,True,GNPDA1
CDH3,True,CDH3
HDAC6,True,HDAC6
PARP2,True,PARP2
MAMLD1,True,MAMLD1
...,...,...
CDC25B,True,CDC25B
OXSR1,True,OXSR1
MVP,True,MVP


In [12]:
# No columns in comparison to our reference structure in adata_sci:
adata_sci.var[list(set(adata_sci.var.columns) - set(adata_l1000.var.columns))]

""
ensembl_id
ENSG00000229258
ENSG00000214702
ENSG00000242855
ENSG00000145103
ENSG00000243770
...
ENSG00000198915
ENSG00000184388
ENSG00000230836


### .X

In [13]:
adata_l1000.X

array([[ 9.4585495,  4.4631004,  8.177425 , ..., 11.85355  , 13.829451 ,
         7.011475 ],
       [12.1437   ,  4.587225 ,  7.6355753, ...,  9.9064245, 13.0807495,
         7.0484   ],
       [ 9.9279   ,  4.55655  ,  9.024925 , ..., 11.3293495, 11.687    ,
         6.92095  ],
       ...,
       [11.2585   ,  9.5756   ,  9.9611   , ..., 10.5406   , 12.3984   ,
         7.51865  ],
       [ 9.5155   ,  7.0375   ,  7.8299   , ..., 10.2312   , 11.6498   ,
         6.8843   ],
       [ 9.30025  ,  5.60155  ,  8.32035  , ..., 12.2567   ,  9.71315  ,
         6.0807   ]], dtype=float32)

In [14]:
adata_sci.X

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 2457641 stored elements and shape (1459, 58302)>

### .layers

In [15]:
adata_l1000.layers

Layers with keys: 

In [16]:
adata_sci.layers

Layers with keys: psbulk_props

## Content

In [17]:
adata_sci.obs.dtypes

plate                      category
well                       category
cell_type                  category
perturbagen                category
pert_type                  category
is_control                 category
pert_dose_uM                float64
pert_time_h                 float64
suspension_type            category
tissue                     category
tissue_type                category
disease                    category
library                    category
stimulation                category
guide                      category
dataset                    category
assay                      category
development_stage          category
organism                   category
sex                        category
self_reported_ethnicity    category
pubchem_cid                category
psbulk_cells                  int64
psbulk_counts                 int64
dtype: object

In [18]:
adata_l1000.obs[list(set(adata_l1000.obs.columns).intersection(set(adata_sci.obs.columns)))].dtypes

pert_type       category
perturbagen     category
psbulk_cells       int64
is_control       boolean
cell_type       category
pert_dose_uM     float64
pert_time_h      float64
plate           category
pubchem_cid     category
well            category
dtype: object

In [19]:
adata_l1000.var.dtypes

symbol             object
pr_gene_symbol     object
pr_is_lm          boolean
dtype: object

In [20]:
adata_l1000.obs[['pert_dose_unit']].value_counts(dropna=False)

pert_dose_unit
uM                1805634
nan                161985
Unknown             23772
uL                   2816
ng                    188
ug/ml                 151
%                      60
Name: count, dtype: int64

In [21]:
adata_l1000.obs[['pert_type']].value_counts(dropna=False)

pert_type  
trt_cp         1805898
ctl_vehicle     102696
ctl_vector       55905
ctl_untrt        30095
ctl_x               12
Name: count, dtype: int64

## Summary:

#### About the general structure:

Standardized l1000 is an anndata object containing obs, var annotations and not sparse count matrix .X. The dictionary-like object layers is empty (I think it could be as l1000 does not have psbulk_props information).

#### Dataset size:
I see that the n_var equals `978`, it is related to the landmark genes, but there are also inferred genes according to the [cmapBQ documentation](https://cmapbq.readthedocs.io/en/latest/cmapBQ.html#module-cmapBQ.query):
+ Best-inferred set of `10,174` genes
+ All inferred genes including `12,328` genes

Are we interested in only landmark genes?

#### Structure of annotations:
+ **.obs** has excessive columns:
  
  `['rna_well', 'pert_itime', 'cmap_name', 'time_val', 'inv_level_10',
       'nearest_dose', 'pert_time_bin', 'project_code', 'perturbation_label',
       'compound_aliases', 'cell_mfc_name', 'dose_bin_uM', 'mw_g_per_mol',
       'pert_id', 'pert_time_unit', 'qc_iqr', 'qc_pass', 'rna_plate',
       'qc_slope', 'sample_id', 'pert_time', 'failure_mode', 'time_unit',
       'inchi_key', 'count_cv', 'qc_f_logp', 'pert_dose_unit', 'count_mean',
       'bead_batch', 'pert_dose', 'dyn_range', 'pert_mfc_id', 'dose_val',
       'build_name', 'alternative_alias', 'pert_idose']`
<br>
  and also lacks of some columns from the standard schema:

  `['psbulk_counts', 'dataset', 'sex', 'library', 'self_reported_ethnicity',
  'tissue', 'guide', 'assay', 'disease', 'stimulation', 'tissue_type',
  'organism', 'suspension_type', 'development_stage']`

  additionally there is a difference in indices of .obs dataframes, in the original schema there is `sample_id` which is constructed as `plate + '_' + well + '_' + perturbagen + '_' + cell_type`, l1000 .obs indices are integers.

+ **.var** has excessive columns:

  `['pr_is_lm', 'pr_gene_symbol']`

    additionally there is a difference in indices of .var dataframes, in the original schema there is `ensembl_id` as an index, and l1000 .var index is `pr_gene_symbol` which corresponds to the gene symbols.

+ **.X** is non-sparse `np.array` which contains `float` values, I think it is reasonable as l1000 does not contain the pure counts.

+ **.layers** is empty, it is reasonable as it might contains just additional information (e.g. produced by `.pp.pseudobulk` function from the `decoupler==2.1.0`).



#### Content of l1000 annotations (based on the original schema):

+ **.obs**

`plate <category>` - ok, id of the plate (e.g. `ABY001_A375_XH_X1_B15`)

`well <category>` - ok, id of the plate (e.g. `D10`)

`cell_type <category>` - sould be **Cellosaurus**, currently `A375`.

`perturbagen <category>` - id of perturbagen (e.g. `943`?, `BRD-A61304759`, `BFP_R1`); we have a name of perturbagen in the original schema.

`pert_type <category>` - indicates the type of perturbagen (e.g. `ctl_vector`, `trt_cp`, `ctl_x`, `ctl_vehicle`, `ctl_untrt`). The original annotation has (e.g. `compound`, `genetic`...). Also, do we consider only compounds and not genetic? In the case we have compounds we need to merge `ctl_vehicle` and `trt_cp` type to `compound` and write (is_contros = True). Here the control compound perturbagens: `['DMSO', 'H2O', 'PBS', 'LIPOFECTAMINE', 'POLYBRENE']`.

`is_control <bool>` - ok, indicates if the perturbation is control or not. (e.g. currently, in l1000 if the perturbagen is in `['DMSO', 'H2O', 'PBS', 'LIPOFECTAMINE', 'POLYBRENE']` - then perturbation is control)

`pert_dose_uM <float64>` - If we consider only compounds as perturbagens then we also need to be able to convert `'ug/ml', '%'` to our `pert_dose_uM`. Currently it is empty. Additionally, I have no idea how to determine uM for `ctl_vector`, `ctl_x`, `ctl_untrt`.

`pert_time_h <float64>` - ok, but I do not know how to convert the info -666.0	into hours.

`suspension_type <category>` - need to be added.

`tissue <category>` - need to be added. Suppose could be taken from `cell_lineage` of `cellinfo` file.


`tissue_type <category>` - need be added.

`disease <category>` - I suppose could be combined as columns `primary_disease` and `subtype` from `cellinfo` file.

`library <category>` - need be added (suppose it should be a column with NaN values).

`stimulation <category>` - need be added (could be a column with NaN values, but I am not sure, need to take a look close).

`guide <category>` - need be added (if we consider compounds I suppose it should be a column with NaN values).

`dataset <category>` - need be added: a name of the dataset, specified manually.

`assay <category>` - need be added (suppose, it could be `L1000 mRNA profiling assay`, but i need to check with the correct efo version).

`development_stage <category>` - need to be added (suppose, it could be derived from `donor_age` column of `cellinfo` file).

`organism <category>` - need to be added (all of cells are from `human` tissues?).

`sex <category>` - need to be added (suppose, it could be derived from `donor_sex` column of `cellinfo` file).

`self_reported_ethnicity <category>` - need to be added (suppose, it could be derived from `donor_ethnicity` column of `cellinfo` file).

`pubchem_cid <category>` - ok. (I have to check the column)

`psbulk_cells <int64>` - it exists, but I am not sure that the value 1 is correct; but also do not know how to derive the correct values. Maybe we need to insert Nans.

`psbulk_counts <int64>` - need to be added: NaN values of the sum of all counts from .X (maybe the second option does not make sense)?

+ **.var**

`symbol <category>` - suppose `pr_gene_symbol` should be renamed to `symbol` and converted to the `<category>`

`ensembl_id` should be added as an index instead of `pr_gene_symbol`.